src/lib/data/attachment.ts

```diff
@@ -0,0 +1,344 @@
+import { isRight } from 'fp-ts/lib/Either';
+import * as t from 'io-ts';
+import { AttachmentId } from '../../db/models/attachment';
+import {
+  AttachmentPrototypeId,
+  AttachmentType,
+} from '../../db/models/attachmentPrototype';
+import { GoverningEntityId } from '../../db/models/governingEntity';
+import { ATTACHMENT_VERSION_VALUE } from '../../db/models/json/attachment';
+import { PlanId } from '../../db/models/plan';
+import { PlanEntityId } from '../../db/models/planEntity';
+import { Database } from '../../db/type';
+import { InstanceDataOfModel } from '../../db/util/raw-model';
+import { organizeObjectsByUniqueProperty } from '../../util';
+import { ioTsErrorFormatter } from '../../util/io-ts';
+import { createBrandedValue } from '../../util/types';
+import { SharedLogContext } from '../logging';
+import { MapOfGoverningEntities } from './governingEntities';
+import {
+  calculateReflectiveTransitiveEntitySupport,
+  ValidatedPlanEntities,
+} from './planEntities';
+
+const ATTACHMENT_DATA = t.union([
+  t.type({
+    type: t.literal('caseLoad'),
+    value: ATTACHMENT_VERSION_VALUE['caseLoad'],
+  }),
+  t.type({
+    type: t.literal('contact'),
+    value: ATTACHMENT_VERSION_VALUE['contact'],
+  }),
+  t.type({
+    type: t.literal('cost'),
+    value: ATTACHMENT_VERSION_VALUE['cost'],
+  }),
+  t.type({
+    type: t.literal('fileWebContent'),
+    value: ATTACHMENT_VERSION_VALUE['fileWebContent'],
+  }),
+  t.type({
+    type: t.literal('indicator'),
+    value: ATTACHMENT_VERSION_VALUE['indicator'],
+  }),
+  t.type({
+    type: t.literal('textWebContent'),
+    value: ATTACHMENT_VERSION_VALUE['textWebContent'],
+  }),
+]);
+
+type PlanAttachmentData = t.TypeOf<typeof ATTACHMENT_DATA>;
+
+type PlanAttachmentParent =
+  | { type: 'plan' }
+  | { type: 'governingEntity'; id: GoverningEntityId }
+  | { type: 'planEntity'; id: PlanEntityId };
+
+export type PlanAttachmentResult = {
+  id: AttachmentId;
+  customRef: string;
+  /**
+   * The part of the plan under which this attachment is directly nested
+   */
+  parent: PlanAttachmentParent;
+  /**
+   * If set, this particular plan attachment is associated with the given
+   * governing entity
+   */
+  governingEntity: GoverningEntityId | null;
+  /**
+   * List of all plan entities that this attachment is associated with,
+   * along with any entities that are transitively supported
+   */
+  supportsEntities: PlanEntityId[];
+  /**
+   * The attachment type, along with the type-checked data from its value
+   */
+  data: PlanAttachmentData;
+};
+
+type PlanAttachmentResults = Map<AttachmentId, PlanAttachmentResult>;
+
+/**
+ * Get all of the attachments of a particular type for a particular plan.
+ *
+ * In addition to fetching the attachments, this also validates the contents of
+ * their value field, ensuring they are of the correct type.
+ */
+export const getAllPlanAttachments = async ({
```

Comment:
This is a third and last time that I talk about how I think we should not mix attachment with plan attachment and attachment of a plan :) There are lots of references like this throughout this file, I won't comment on each
```suggestion
export const getAllAttachmentsForPlan = async ({
```